dmriprep/utils/grouping.py

```diff
@@ -0,0 +1,711 @@
+
+    Some missing metadata
+    >>> dwi_files = [
+    ...     '/data/sub-1/dwi/sub-1_run-1_dwi.nii.gz',
+    ...     '/data/sub-1/dwi/sub-1_run-2_dwi.nii.gz',
+    ...     '/data/sub-1/dwi/sub-1_run-3_dwi.nii.gz',
+    ...     '/data/sub-1/dwi/sub-1_run-4_dwi.nii.gz',
+    ...     '/data/sub-1/dwi/sub-1_run-5_dwi.nii.gz'
+    ... ]
+    >>> metadatas = [
+    ...     {'PhaseEncodingDirection': 'j'},
+    ...     {'PhaseEncodingDirection': 'j'},
+    ...     {'PhaseEncodingDirection': 'j-'},
+    ...     {},
+    ...     {}
+    ... ]
+    >>> split_by_phase_encoding_direction(dwi_files, metadatas) # doctest: +NORMALIZE_WHITESPACE
+    [{'dwi_series': ['/data/sub-1/dwi/sub-1_run-1_dwi.nii.gz',
+       '/data/sub-1/dwi/sub-1_run-2_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': None},
+      'dwi_series_pedir': 'j',
+      'concatenated_bids_name': 'sub-1'},
+     {'dwi_series': ['/data/sub-1/dwi/sub-1_run-3_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': None},
+      'dwi_series_pedir': 'j-',
+      'concatenated_bids_name': 'sub-1_run-3'},
+     {'dwi_series': ['/data/sub-1/dwi/sub-1_run-4_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': None},
+      'dwi_series_pedir': '',
+      'concatenated_bids_name': 'sub-1_run-4'},
+     {'dwi_series': ['/data/sub-1/dwi/sub-1_run-5_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': None},
+      'dwi_series_pedir': '',
+      'concatenated_bids_name': 'sub-1_run-5'}]
+
+
+
+    """
+    pe_dir_groups = defaultdict(list)
+    unknowns = []
+    for dwi_file, meta in zip(dwi_files, metadatas):
+        pe_dir = meta.get("PhaseEncodingDirection")
+        if pe_dir:
+            pe_dir_groups[pe_dir].append(dwi_file)
+        else:
+            unknowns.append(dwi_file)
+
+    dwi_groups = []
+    for pe_dir, dwi_group in sorted(pe_dir_groups.items()):
+        dwi_groups.append(
+            {'dwi_series': dwi_group,
+             'fieldmap_info': {'suffix': None},
+             'dwi_series_pedir': pe_dir,
+             'concatenated_bids_name': get_concatenated_bids_name(dwi_group)})
+    for unknown in unknowns:
+                dwi_groups.append(
+                    {'dwi_series': [unknown],
+                     'fieldmap_info': {'suffix': None},
+                     'dwi_series_pedir': '',
+                     'concatenated_bids_name': get_concatenated_bids_name([unknown])})
+    return dwi_groups
+
+
+def group_by_warpspace(dwi_files, layout, ignore_fieldmaps):
+    """Groups a session's DWI files by their acquisition parameters.
+
+    DWIs are grouped by their **warped space**. Two DWI series that are
+    listed in the IntendedFor field of a fieldmap are assumed to have the same
+    susceptibility distortions and therefore be in the same warped space. The goal
+    of this function is to combine DWI series into groups of acquisitions that
+    are in the same warped space into a list of scans that can be combined after
+    unwarping.
+
+    Parameters
+    -----------
+
```

Comment:
There shouldn't be any empty line after the section header (same for all other instances)